starlette/middleware/http.py

```diff
@@ -0,0 +1,94 @@
+from contextlib import aclosing
+from functools import partial
+from typing import AsyncGenerator, Callable, Optional
+
+from ..datastructures import MutableHeaders
+from ..responses import Response
+from ..types import ASGIApp, Message, Receive, Scope, Send
+
+HTTPDispatchFlow = AsyncGenerator[Optional[Response], Response]
+
+
+class HTTPMiddleware:
+    def __init__(
+        self,
+        app: ASGIApp,
+        dispatch_func: Optional[Callable[[Scope], HTTPDispatchFlow]] = None,
+    ) -> None:
+        self.app = app
+        self.dispatch_func = self.dispatch if dispatch_func is None else dispatch_func
+
+    def dispatch(self, scope: Scope) -> HTTPDispatchFlow:
+        raise NotImplementedError  # pragma: no cover
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        async with aclosing(self.dispatch(scope)) as flow:
+            # Kick the flow until the first `yield`.
+            # Might respond early before we call into the app.
+            early_response = await flow.__anext__()
+
+            if early_response is not None:
+                await early_response(scope, receive, send)
```

Comment:
```suggestion
            early_response_option = await flow.__anext__()

            if early_response_option is not None:
                await early_response_option(scope, receive, send)
```

Reply:
Not sure what prompts this? We know from type hints that this is an `Response | None`, I don't think we need to add its "optional" nature in the name.

Reply:
Yeah you don't have to accept it. I got a little bit confused reading it without type hints because I thought that maybe they use her was returning an ASGI app (well they are, it's a Response, I thought an arbitrary ASGI callable). I also tend to use `optional_...` or `maybe_...`, but that's just my preference